### DISCUSSION TITLE
may Publisher be the right tool for me?

### DISCUSSION THREAD

**Discussion by pr-apes** [speedata/publisher#374] [General]
@pgundlach,

I have some templates in ConTeXt (written about five years ago [or so I was told]).

These help to generate different PDF documents (letters, invoices, envelopes and even printed return receipts), used by other people.

Although the system is pretty efficient in solving the document generation needs, it is a pain for me to extend (or even modify) it.

My have extremely basic coding abilities (no actual programming knowledge) and very limited exposure to XML, ConTeXt and _Lua_.

I was told that your _Publisher_ could be a tool for this task.

I must confess that _Publisher_ looks pretty amazing and the examples look gorgeous.

I wonder whether _Publisher_ may fit our needs.

Our main requirements are very strict, mainly because of workflows (which cannot be changed):

* Mail merging capabilities reading data for template from a CSV file.

* All different document classes should be generated from a single template.

* Mail merging should generate single PDF files (not imposed from a single PDF document).

Before starting to learn _Publisher_ I need to summarize the requirements for this project.

The above list doesn’t contain all requirements (I have to elaborate further), but I think these ones are a must.

As a first question to decide whether I should go for it (or use another tool),  will _Publisher_ be the right tool for me?

Many thanks for your help.

**Comment by pgundlach** [comment]
The documents you list (letters, invoices, envelopes and even printed return receipts) looks doable to me. If you were asking about more text like documents like magazines, scientific papers or such, I'd say TeX is more suitable.

Your third point is not possible with the speedata Publisher, but it may be possible to create n separate data files and one publishing run for each data file. 

That said, I suggest you try to create a document and if you have questions, don't hesitate to ask.

I cannot answer the question if the Publisher is the right tool for you, but I can help you to find out.

**Reply by pr-apes** [reply]
> I cannot answer the question if the Publisher is the right tool for you, but I can help you to find out.

Many thanks for your reply and your help, @pgundlach.

Of course, I have to decide whether _Publisher_ is the right tool for the task that I have to complete 😅.

> The documents you list (letters, invoices, envelopes and even printed return receipts) looks doable to me. If you were asking about more text like documents like magazines, scientific papers or such, I'd say TeX is more suitable.

I realize that _Publisher_ isn’t intended as a full replacement for TeX (neither do I understand it as such).

**Comment by pr-apes** [comment]
I have seen that there that CSV files may be read (and converted to XML, https://doc.speedata.de/publisher/en/advancedtopics/preprocessing/#_read_csv_files).

The ability to generate single documents is essential for us.

Of course, I have to learn how to _Publisher_ myself.

But I need to know whether it is going to fit.

Imagine the following CSV data:

``` csv
"Mary";"Literature"
;"Chemistry"
;"French"
"Anne";"Physics"
;"Math"
"John";"German"
;"Greek"
;"Environmental Sciences"
"Jane";"Chemistry"
;"French"
;"Physics"
;"Math"
;"German"
```

The following XML source:

``` xml
<p><variable:c1/> has the following subjects: <variable:c2/></p>
```

And the layout:

``` xml
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

<!-- <Options mainlanguage="en" /> not required, luatex defaults this -->
  <Trace hyphenation="yes" />
<LoadFontfile name="CrimsonPro-Regular" filename="CrimsonPro-Regular.ttf" />

<DefineFontfamily name="textfont" fontsize="9" leading="11">
  <Regular fontface="CrimsonPro-Regular"/>
</DefineFontfamily>

<Pageformat width="100mm" height="50mm"/>
	<Record element="p">
	  <PlaceObject>
		<Textblock fontfamily="textfont">
		  <Paragraph>
			<Value select="."/>
		  </Paragraph>
		</Textblock>
	  </PlaceObject>
	</Record>
</Layout>
```

I don’t know how to get individual PDF documents (only by merging CSV and XML data), containing each the following content:

1. Mary has the following subjects: Literature, Chemistry and French.

1. Anne has the following subjects: Physics and Math.

1. John has the following subjects: German, Greek and Environmental Sciences.

1. Jane has the following subjects: Chemistry, French, Physics, Math and German.

This is the most basic way of sampling what I need to achieve.

Many thanks for your help.

**Reply by pgundlach** [reply]
I think you should write some kind of pre-processing script (Ruby, Python, ...) to create a data file for each of these lines from CSV. The built-in CSV does not work, because it is already part of a publishing run.

You could of course run the speedata Publisher with a variable (`sp --var person=Mary` or `sp --var line=1`) and use the built in Lua script to filter the CSV. Then it would be the task to of the calling application to create the correct sp calls.

**Reply by pr-apes** [reply]
> I think you should write some kind of pre-processing script (Ruby, Python, ...) to create a data file for each of these lines from CSV. The built-in CSV does not work, because it is already part of a publishing run.

I have to avoid adding extra dependencies.

> You could of course run the speedata Publisher with a variable (`sp --var person=Mary` or `sp --var line=1`) and use the built in Lua script to filter the CSV. Then it would be the task to of the calling application to create the correct `sp` calls.

And how about a _Lua_ filter that makes these calls with `sp --var non_empty_column=1` and loops the whole CSV file?

**Comment by pgundlach** [comment]
@pr-apes currently there is no way to start another sp process from the Lua filter, so the last idea does not at the moment.

**Comment by pgundlach** [comment]
Just in case you're still interested, the speedata Publisher can now run external programs from the Lua preprocessing state (version 4.9.7, command is `runtime.execute()`).

**Reply by pr-apes** [reply]
This is really great news!

**Reply by pr-apes** [reply]
Many thanks for the improvement!
